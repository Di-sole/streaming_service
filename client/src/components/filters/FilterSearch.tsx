import { useEffect, useState } from 'react';
import styles from './FilterSearch.module.scss';
import Autosuggest from 'react-autosuggest';
import axios, { AxiosResponse } from 'axios';
import theme from './theme.module.scss';
import { useTranslation } from 'next-i18next';
import IPerson from '../../models/IPerson';
import { useRouter } from 'next/router';
import { useAppDispatch, useAppSelector } from '@/store/hooks/redux';
import { addFilteredMovies } from '@/store/slices/moviesSlice';

interface ISearch {
    category: 'director' | 'actor';
    //searchBy: string;
}

const FilterSearch = ({ category }: ISearch) => {
    const { t } = useTranslation('moviesPage');
    const dispatch = useAppDispatch();
    const [inputValue, setInputValue] = useState('');
    const [suggestions, setSuggestions] = useState<string[]>([]);
    const [searchedPerson, setSearchedPerson] = useState('');

    const searchParam = category === 'director' ? t('filterPanel.searchByDirector') : t('filterPanel.searchByActor');

    const actors = useAppSelector((state) => state.actors.actors);

    const handleEnterPressed = ({ target, key }: KeyboardEvent): void => {
        const searchInput = target as HTMLInputElement;
        if (key === 'Enter') {
            setSearchedPerson(searchInput!.value);
        }
    };

    useEffect(() => {
        const fetchActorMovies = async () => {
            try {
                const response = await axios.get(`http://localhost:6125/movies?actor=${searchedPerson}`);
                console.log('actorMovies:', response.data.docs[0].page);
                //     const names: string[] = result.data
                //         .map((person: IPerson) => person.name)
                //         .filter((person: string[]) => person && person.includes(inputValue));
                dispatch(addFilteredMovies(response.data.docs[0].page));
            } catch (err) {
                console.log(err);
            }
        };
        fetchActorMovies();
    }, [searchedPerson]);

    return (
        <div className={styles.container__formContent}>
            <div className={styles.container__inputBody} onKeyDown={handleEnterPressed as VoidFunction}>
                <Autosuggest
                    theme={theme}
                    inputProps={{
                        placeholder: searchParam,
                        name: 'person',
                        value: inputValue,
                        onChange: (_event, { newValue }) => {
                            setInputValue(newValue);
                        },
                    }}
                    suggestions={suggestions}
                    onSuggestionsFetchRequested={({ value }) => {
                        if (!value) {
                            setSuggestions([]);
                            return;
                        }
                        const names: string[] = actors
                            .map((person: IPerson) => person.name)
                            .filter((person: string) => person && person.includes(inputValue));
                        setSuggestions(names);
                        // try {
                        //     const query = searchBy === 'Актер' ? 'getAllActors' : 'getAllDirectors';
                        //     const result = await axios.get(`http://localhost:6125/${query}`);
                        //     const names: string[] = result.data
                        //         .map((person: IPerson) => person.name)
                        //         .filter((person: string[]) => person && person.includes(inputValue));
                        //     setSuggestions(names);
                        // } catch (error) {
                        //     console.log(error);
                        // }
                    }}
                    onSuggestionsClearRequested={() => {
                        setSuggestions([]);
                    }}
                    onSuggestionSelected={(event, { suggestion, method }) => {
                        const target = event.target as HTMLInputElement;
                        if (method === 'enter') {
                            setSearchedPerson(target.value);
                        }
                    }}
                    getSuggestionValue={(suggestion) => suggestion}
                    renderSuggestion={(suggestion) => <div>{suggestion}</div>}
                />
                <div className={styles.container__fieldButton}>
                    <div
                        className={[
                            styles.container__buttonIcon,
                            inputValue === ''
                                ? styles.container__buttonIcon_search
                                : styles.container__buttonIcon_remove,
                        ].join(' ')}
                        onClick={() => setInputValue('')}
                    ></div>
                </div>
            </div>
        </div>
    );
};

export default FilterSearch;
