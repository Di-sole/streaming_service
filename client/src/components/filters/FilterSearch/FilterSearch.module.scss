.container {
    padding-top: 12px;

    &__formContent * {
        transition-duration: 0.4s;
    }

    &__inputBody {
        position: relative;
        display: flex;
        margin-left: -10px;
        margin-right: -10px;
        overflow: hidden;
        border-radius: 8px;
    }

    &__fieldButton {
        position: absolute;
        right: 0;

        height: 40px;
        width: 40px;

        opacity: 0.3;

        &:hover {
            opacity: 1;
        }
    }

    &__buttonIcon {
        position: absolute;
        top: 12px;
        bottom: 12px;
        left: 12px;
        right: 12px;

        &_search {
            pointer-events: none;
            &::before {
                content: url('../../../../public/icons/search.svg');
                opacity: 0.3;
                transition: color 0.4s;
            }
        }

        &_remove {
            cursor: pointer;
            pointer-events: all;
            &::before {
                content: url('../../../../public/icons/close.svg');
                opacity: 0.3;
                transition: color 0.4s;
            }
        }
    }
}
