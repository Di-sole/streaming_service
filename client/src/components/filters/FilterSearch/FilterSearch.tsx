import { useEffect, useState } from 'react';
import styles from './FilterSearch.module.scss';
import Autosuggest from 'react-autosuggest';
import axios from 'axios';
import theme from '../theme.module.scss';
import { useTranslation } from 'next-i18next';
import IPerson from '../../../models/IPerson';
import { useAppDispatch, useAppSelector } from '../../../store/hooks/redux';
import { actorFilterAdded, addFilteredMovies, directorFilterAdded } from '../../../store/slices/moviesSlice';
import { checkFiltersStatus } from '../../../utils/functions';

interface ISearch {
    suggestionsList: IPerson[];
    category: 'director' | 'actor';
}

const FilterSearch = ({ suggestionsList, category }: ISearch) => {
    const { t } = useTranslation('moviesPage');
    const dispatch = useAppDispatch();
    const [inputValue, setInputValue] = useState('');
    const [suggestions, setSuggestions] = useState<string[]>([]);
    const [searchedPerson, setSearchedPerson] = useState('');

    const personNames: string[] = [];
    suggestionsList.forEach((item) => {
        personNames.push(item.name);
        personNames.push(item.enName);
    });

    const allFilters = useAppSelector((state) => state.movies.filters);
    const joinedQuery = `actor=${allFilters.actor}&director=${allFilters.director}`;

    useEffect(() => {
        category === 'actor' ? dispatch(actorFilterAdded(inputValue)) : dispatch(directorFilterAdded(inputValue));
    }, [inputValue]);

    const searchParam = category === 'director' ? t('filterPanel.searchByDirector') : t('filterPanel.searchByActor');

    const handleEnterPressed = ({ target, key }: KeyboardEvent): void => {
        const searchInput = target as HTMLInputElement;
        if (key === 'Enter') {
            setSearchedPerson(searchInput!.value);
        }
    };

    useEffect(() => {
        const fetchActorMovies = async () => {
            if (searchedPerson.length === 0) {
                return;
            }
            try {
                const response = await axios.get(`http://localhost:6125/movies?${joinedQuery}`);
                dispatch(addFilteredMovies(response.data.docs[0].page));
            } catch (e: any) {
                console.log(e.response?.data?.message);
            }
        };
        fetchActorMovies();
    }, [searchedPerson]);

    useEffect(() => {
        if (checkFiltersStatus(allFilters)) {
            setSearchedPerson('');
        }
    }, [allFilters]);

    return (
        <div className={styles.container__formContent}>
            <div className={styles.container__inputBody} onKeyDown={handleEnterPressed as VoidFunction}>
                <Autosuggest
                    theme={theme}
                    inputProps={{
                        placeholder: searchParam,
                        name: 'person',
                        value: inputValue,
                        onChange: (_event, { newValue }) => {
                            setInputValue(newValue);
                        },
                    }}
                    suggestions={suggestions}
                    onSuggestionsFetchRequested={({ value }) => {
                        if (!value) {
                            setSuggestions([]);
                            return;
                        }
                        const names: string[] = personNames.filter(
                            (person: string) => person && person.toLowerCase().includes(inputValue.toLowerCase()),
                        );
                        setSuggestions(names);
                    }}
                    onSuggestionsClearRequested={() => {
                        setSuggestions([]);
                    }}
                    onSuggestionSelected={(event, { suggestion, method }) => {
                        const target = event.target as HTMLInputElement;
                        if (method === 'enter') {
                            setSearchedPerson(target.value);
                        }
                    }}
                    getSuggestionValue={(suggestion) => suggestion}
                    renderSuggestion={(suggestion) => <div>{suggestion}</div>}
                />
                <div className={styles.container__fieldButton}>
                    <div
                        className={[
                            styles.container__buttonIcon,
                            inputValue === ''
                                ? styles.container__buttonIcon_search
                                : styles.container__buttonIcon_remove,
                        ].join(' ')}
                        onClick={() => setInputValue('')}
                    ></div>
                </div>
            </div>
        </div>
    );
};

export default FilterSearch;
