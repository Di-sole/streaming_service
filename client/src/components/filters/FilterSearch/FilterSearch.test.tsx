import { render, screen } from '@testing-library/react';
import FilterSearch from './FilterSearch';
import * as reduxHooks from '../../../store/hooks/redux';
import { actorSuggestions } from '../../../testAsserts/testItems';

jest.mock('../../../store/hooks/redux');
const mockedSelector = jest.spyOn(reduxHooks, 'useAppSelector');
const useDispatch = jest.spyOn(reduxHooks, 'useAppDispatch');
const mockDispatch = jest.fn();
const filters = {
    genres: [],
    countries: [],
    ratingKp: 0,
    votesKp: 0,
    director: '',
    actor: '',
};

jest.mock('next-i18next', () => ({
    useTranslation: () => {
        return {
            t: (str: string) => str,
            i18n: {
                changeLanguage: () => new Promise(() => {}),
            },
        };
    },
}));

describe('FILTERSEARCH SNAPSHOTS TESTS', () => {
    test('"category" director FilterSearch should not be changed', () => {
        mockedSelector.mockReturnValue(filters);
        useDispatch.mockReturnValue(mockDispatch);
        render(<FilterSearch category="director" suggestionsList={actorSuggestions} />);

        const div = screen.getByTestId('filterSearch');
        expect(div).toMatchSnapshot();
    });

    test('"category" actor FilterSearch should not be changed', () => {
        mockedSelector.mockReturnValue(filters);
        useDispatch.mockReturnValue(mockDispatch);
        render(<FilterSearch category="actor" suggestionsList={actorSuggestions} />);

        const div = screen.getByTestId('filterSearch');
        expect(div).toMatchSnapshot();
    });
});
