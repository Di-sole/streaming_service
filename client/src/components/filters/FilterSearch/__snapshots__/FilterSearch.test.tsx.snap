// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FILTERSEARCH SNAPSHOTS TESTS "category" actor FilterSearch should not be changed 1`] = `
<div
  class="container__formContent"
  data-testid="filterSearch"
>
  <div
    class="container__inputBody"
  >
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-owns="react-autowhatever-1"
      role="combobox"
    >
      <input
        aria-autocomplete="list"
        aria-controls="react-autowhatever-1"
        autocomplete="off"
        name="person"
        placeholder="filterPanel.searchByActor"
        type="text"
        value=""
      />
      <div
        id="react-autowhatever-1"
        role="listbox"
      />
    </div>
    <div
      class="container__fieldButton"
    >
      <div
        class="container__buttonIcon container__buttonIcon_search"
      />
    </div>
  </div>
</div>
`;

exports[`FILTERSEARCH SNAPSHOTS TESTS "category" director FilterSearch should not be changed 1`] = `
<div
  class="container__formContent"
  data-testid="filterSearch"
>
  <div
    class="container__inputBody"
  >
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-owns="react-autowhatever-1"
      role="combobox"
    >
      <input
        aria-autocomplete="list"
        aria-controls="react-autowhatever-1"
        autocomplete="off"
        name="person"
        placeholder="filterPanel.searchByDirector"
        type="text"
        value=""
      />
      <div
        id="react-autowhatever-1"
        role="listbox"
      />
    </div>
    <div
      class="container__fieldButton"
    >
      <div
        class="container__buttonIcon container__buttonIcon_search"
      />
    </div>
  </div>
</div>
`;
